### PR TITLE
Add recompute= keyword to svd_compressed for lower-memory use

### PR DESCRIPTION
**Pull request by mrocklin**
The svd_compressed algorithm needs to use the input array a few times
Often, this array is too large to fit into memory, and so it is better
to recompute it.  This PR adds a keyword, ``recompute=``, which
optionally triggers this behavior.  It is turned off by default.

We also fix up the docstring a bit and one of the keys in the high level
graph.

This also introduces a dask.distributed agnostic wait function to wait
on persisted data if we're using the distributed scheduler, but to do
nothing otherwise.

With this change we're able to compute arbitrarily large svd
computations in low memory.

cc @alimanfoo

### CODE REVIEW

**Comment by mrocklin** [comment]
I'm able to compute the first few components of the `v` result of an SVD of an array of size (n, 100,000), for any `n` on my laptop in around 4 GB of RAM.

https://gist.github.com/mrocklin/b7f3f9fe68d4f6c49a1470cd9dcfc2be

File: dask/array/tests/test_linalg.py [hunk 1]
```diff
@@ -910,3 +910,13 @@ def test_norm_implemented_errors(shape, chunks, axis, norm, keepdims):
     if len(shape) > 2 and len(axis) == 2:
         with pytest.raises(NotImplementedError):
             da.linalg.norm(d, ord=norm, axis=axis, keepdims=keepdims)
+
+
+def test_biggish_svd():
```
Comment (TomAugspurger) [inline dask/array/tests/test_linalg.py]: How long does this take? Mark with `slow`?

File: dask/array/linalg.py [hunk 2]
```diff
@@ -641,6 +642,10 @@ def compression_matrix(data, q, n_power_iter=0, seed=None):
     n_power_iter: int
         number of power iterations, useful when the singular values of
         the input matrix decay very slowly.
+    recompute : bool
+        Should we compute data at each stage during computation?
```
Comment (TomAugspurger) [inline dask/array/linalg.py]: Stage -> power iteration? I'm not sure what the common term is here, but this links it strongly with the previous parameter.

**Comment by mrocklin** [comment]
cc @pentschev, in case you're interested in trying this out on a multi-GPU machine to test performance

File: dask/array/linalg.py [hunk 3]
```diff
@@ -660,11 +665,17 @@ def compression_matrix(data, q, n_power_iter=0, seed=None):
     mat_h = data.dot(omega)
     for j in range(n_power_iter):
         mat_h = data.dot(data.T.dot(mat_h))
+        if recompute:
+            mat_h = mat_h.persist()
+            wait(mat_h)
     q, _ = tsqr(mat_h)
+    if recompute:
```
Comment (TomAugspurger) [inline dask/array/linalg.py]: Just from reading the docstring, I can't decide whether or not I would `recompute=True` would imply a `wait` here. Can you give the motivation for it?

File: dask/array/tests/test_linalg.py (hunk 1 above)
Comment (mrocklin) [reply dask/array/tests/test_linalg.py]: Not long.  I've renamed the test.

File: dask/array/linalg.py (hunk 2 above)
Comment (mrocklin) [reply dask/array/linalg.py]: Yes, each power iteration, and also the first matmul/tsqr call (which is possibly more important).  I'm not sure that we can call this each power iteration.

**Comment by pentschev** [comment]
@mrocklin does this reduce compute time on CPU too, or just memory usage?

**Comment by mrocklin** [comment]
Assuming that we're computing the input to the `svd_compressed` call then it increases the computation that we have to do because we recompute the input repeatedly.  However, for larger-than-memory inputs this may decrease overall time because we won't be spilling to disk.

**Comment by pentschev** [comment]
I see. That's a bit more difficult to benchmark accurately, since the involved factors are many (when benchmarking against a spilling to host or disk case). I don't think I've done any realistic large memory benchmarks like this, I will have to think of a somewhat reliable way of benchmarking all this.

**Comment by pentschev** [comment]
Nevertheless, it's great to have this option! Thanks for doing that @mrocklin.

**Comment by mrocklin** [comment]
To be clear, the objective here is to avoid spilling to disk.  Previously we would have had to, now we can recompute the input.  My *guess* here is that on CPU we'll be bound by in-memory computations like matrix-matrix multiplies and QR decompositions while with a couple GPUs in the mix we'll be more bound by disk read speeds.

**Comment by pentschev** [comment]
> To be clear, the objective here is to avoid spilling to disk.

Yes, it's very clear. :)

> My guess here is that on CPU we'll be bound by in-memory computations like matrix-matrix multiplies and QR decompositions while with a couple GPUs in the mix we'll be more bound by disk read speeds.

You may be right, but we have to consider where data really lies, is it still on disk mostly, or has it been loaded to the main memory already? We'll have to think carefully how to benchmark this correctly, it's easy enough to screw up simple benchmarks, imagine this one.

**Comment by mrocklin** [comment]
> or has it been loaded to the main memory already?

Through @alimanfoo I've become interested in a case where it won't easily fit in main memory.  I'm imagining the situation where it's several times too large.

File: dask/array/linalg.py (hunk 3 above)
Comment (mrocklin) [reply dask/array/linalg.py]: We need to remove `data` from the graph of `q` before `data` gets used again in `svd_compressed`.  However, I see now that this is only necessary in the context of `svd_compressed`, so I've moved it there.

**Comment by mrocklin** [comment]
(@pentschev I'm also not suggesting this as a high priority of something for you to work on, I just thought that you might find it interesting)

**Comment by pentschev** [comment]
Yep, I understand that. I was just thinking out loud here and raising the question of how can that be done correctly and fairly to benchmark the entire hardware/software stack.

File: dask/base.py [hunk 4]
```diff
@@ -1011,3 +1011,18 @@ def get_scheduler(get=None, scheduler=None, collections=None, cls=None):
...
+    This is a compatibility alias for ``dask.distributed.wait``.
+    If it is applied onto Dask collections without Dask Futures or if Dask
+    distributed is not installed then it is a no-op
+    """
+    try:
+        from distributed import wait
+
+        return wait(x, timeout=timeout, return_when=return_when)
+    except (ImportError, ValueError):
+        return x
```
Comment (mrocklin) [inline dask/base.py]: @jcrist any objections with this?

**Comment by mrocklin** [comment]
@alimanfoo can you remind me of the code you used to set up a random and renormalized array that looked like the data you had?

**Comment by alimanfoo** [comment]
> @alimanfoo can you remind me of the code you used to set up a random and renormalized array that looked like the data you had?

Thanks @mrocklin for this, very cool. Here's the notebook we were using: https://gist.github.com/alimanfoo/5c9c546b42550d0b89d647a1b55ac134

File: dask/array/linalg.py (hunk 2 above)
Comment (jcrist) [reply dask/array/linalg.py]: How about each use? Non-technical term, and leaves slack in definition if the implementation changes.

File: dask/base.py (hunk 4 above)
Comment (jcrist) [reply dask/base.py]: This seems fine to me. I'm hesitant to make this a public function (seems useful for our internals, but wouldn't necessarily want users to start relying on it).  Leaving it out of top-level imports makes sense.

File: dask/array/linalg.py [hunk 5]
```diff
@@ -250,13 +251,13 @@ def tsqr(data, compute_svd=False, _max_vchunk_size=None):
         dependencies[name_q_st2] = q_inner.__dask_layers__()
 
         # R: R_inner
-        name_r_st2 = "r-inner-" + token
+        name_r_st2 = "r-inner" + token
```
Comment (jcrist) [inline dask/array/linalg.py]: Why remove the dash before the token?

Comment (mrocklin) [reply dask/array/linalg.py]: The `token` variable has a dash in it already, so we were getting double-dashes.

**Pull request merged**